new setupcfg writer

clavedeluna opened the pull request:
## Overview
*Adds a setup.cfg file writer for new dependencies*

## Description

* This writer relies on `configparser` only to detect the existing dependencies in the setup.cfg file. We can't rely on it to update and write to file because `configparser` doesn't retain some formatting and comment lines.
* I tried some other parsers to see if they could work, but none did. Instead, this is an attempt to manually write the new dependencies to file with some formatting heuristics and as much reasonable error catching as I could think.

clavedeluna committed:
new setupcfg writer

clavedeluna commented:
I made this PR ready to review despite the codecov failure which is valid but due to my attempt to catch exceptions. After review I"ll go back and add some testing or skip coverage in certain lines.

drdavella reviewed (commented):
Just curious but did you investigate whether this project would help simplify the implementation? https://github.com/Preocts/commented-configparser

clavedeluna commented:
> Just curious but did you investigate whether this project would help simplify the implementation? https://github.com/Preocts/commented-configparser

I tested with commented-configparser and am very slightly disappointed that it's only a tiny bit buggy. It does retain the comments in most cases, but in some cases it moves comments to different lines. It also added some trailing spaces to the new setup.cfg file. So here's a diff:

```diff
 #comment
 [metadata] 
name = my_package
 version = attr: my_package.VERSION
 author = Josiah Carberry
 author_email = josiah_carberry@brown.edu
 description = My package description
 long_description = file: README.rst, CHANGELOG.rst, LICENSE.rst
 keywords = one, two
 # comment should stay in ln 10
 license = BSD-3-Clause
-classifiers =
+classifiers =
+# comment should stay in ln 16
        Framework :: Django
        Programming Language :: Python :: 3

-# comment should stay in ln 16
```
changes are:
1. added a leading spaces after classifiers =
2. moved comment in ln 16 to ln13, while not changing comment in ln1

@drdavella lmk if you want to accept these changes and I can switch to using commented config or we can keep code as is

drdavella commented:
@clavedeluna thanks for the info. I'm sorry it didn't work out of the box. Let's accept this as-is for now and we can revisit later since I don't think this is the most critical work for us right now. One other possibility down the road would be to vendor in that third-party code (assuming a permissive license) and maybe tweak it to get the behavior we need.

drdavella commented on src/codemodder/dependency_management/setupcfg_writer.py:
```diff
@@ -0,0 +1,124 @@
+        configparser does not retain formatting or comment lines, so we have to build
+        the output newline manually.
+        """
+        clean_lines = list(map(lambda s: s.strip(), original_lines))
```
I like `map` but genuinely curious whether there's any benefit to doing it this way rather than using a comprehension?
```python
clean_lines = [s.strip for s in original_lines]
```

clavedeluna replied:
no other benefit to map, I think the moment I wrote it that's the first thing that came to me. I'll switch it out

drdavella commented on tests/dependency_management/test_setupcfgt_writer.py:
```diff
@@ -0,0 +1,302 @@
+    assert setup_cfg.read() == dedent(updated_setupcfg)
+
+
+def test_dont_add_existing_dependency(tmpdir):
```
We should have another test case where the version of `security` that already exists in the package is different from the one being added. There should be no new dependency added (or version changed) in this case either.

clavedeluna replied:
This is a really good test case to think about. None of the other writers handle for this either so if it's ok with you it would be good to handle it in a separate PR for all. It would be worth thinking about it a bit more:  do we really not want to update the version? What if the project has `pkg==1.0` but our recommendation is `pkg==2.0`?

Worth thinking about it independently.

drdavella reviewed (approved):
LGTM but it would be good to add the test case mentioned below.

clavedeluna committed:
add test coverage for setup.cfg
